src/simulation/elements/SLCN.cpp

```diff
@@ -0,0 +1,236 @@
+
+static float clamp(float n, float low, float high) {
+  return std::max(low, std::min(n, high));
+}
+
+static int update(UPDATE_FUNC_ARGS)
+{
+	Particle &self = parts[i];
+
+	float velocity = std::abs(std::sqrt(std::pow(self.vx, 2) + std::pow(self.vy, 2)));
+
+	//FROM: GOLD.cpp:74 but prettied up a bit.
+	if (self.life == 0 && self.temp < 373.15f)
+	{
+		for (int j = 0; j < 4; j++)
+		{ 
+			int check_coords_x[] = { -4, 4, 0, 0 };
+			int check_coords_y[] = { 0, 0, -4, 4 };
+
+			int neighbour_x = check_coords_x[j];
+			int neighbour_y = check_coords_y[j];
+
+			int neighbour_id = pmap[y + neighbour_y][x + neighbour_x];
+
+			if (neighbour_id == 0) continue;
+
+			int neighbour_type = TYP(neighbour_id);
+			Particle &neighbour = parts[ID(neighbour_id)];
+
+			if (neighbour_type == PT_SPRK && neighbour.life != 0 && neighbour.life < 4)
+			{
+				sim->part_change_type(i, x, y, PT_SPRK);
+				self.life = 4;
+				self.ctype = PT_SLCN;
+			}
+		}
+	}
+
+	int gt_phase_count = 0;
+	
+	for (int neighbour_x=-2; neighbour_x<3; neighbour_x++)
+		for (int neighbour_y=-2; neighbour_y<3; neighbour_y++) 
+		{
+			int neighbour_id = pmap[y + neighbour_y][x + neighbour_x];
+
+			if (neighbour_id == 0) continue;
+
+			int neighbour_type = TYP(neighbour_id);
+			Particle &neighbour = parts[ID(neighbour_id)];
+			if (neighbour_type == PT_SLCN)
+			{
+				if (self.tmp == neighbour.tmp)
+				{
+					gt_phase_count += 1;
+				}
+			}
+		}
+
+	// Phase change logic.
+	int current_color_phase = self.tmp & 31; // limited to 0 through 31. using bitlogic for simplicity.
+	int current_spark_phase = self.tmp & 15; // limited to 0 through 15.
+
+	float &clr_phse_transition_cnt = self.pavg[0];
+	float &sprk_phse_transition_cnt = self.pavg[1];
+
+	// prevent these values from going through the roof.
+	clr_phse_transition_cnt = clamp(clr_phse_transition_cnt, 0.0f, PHASE_THRESHOLD * 2);
+	sprk_phse_transition_cnt = clamp(sprk_phse_transition_cnt, 0.0f, PHASE_THRESHOLD * 2);
+
+	clr_phse_transition_cnt += (velocity * VELOCITY_MULTIPLIER) + SPARKLE_RATE + std::abs(RNG::Ref().uniform01() * SPARKLE_RATE);
+	sprk_phse_transition_cnt += (velocity * VELOCITY_MULTIPLIER) + SPARKLE_RATE + std::abs(RNG::Ref().uniform01() * SPARKLE_RATE);
+
+	// evil trick to combat syncronization.
+	if (gt_phase_count > 3) {
+		if (RNG::Ref().chance(1, 30)) {
+			self.tmp = RNG::Ref().between(0, 31);
+			self.tmp2 = RNG::Ref().between(0, 15);
+			clr_phse_transition_cnt = (float)RNG::Ref().between(0, (int)PHASE_THRESHOLD);
+			sprk_phse_transition_cnt = (float)RNG::Ref().between(0, (int)PHASE_THRESHOLD);
+		}
+	}
+
+	if (clr_phse_transition_cnt > PHASE_THRESHOLD) {
+		clr_phse_transition_cnt -= PHASE_THRESHOLD;
+		self.tmp = (current_color_phase + 1) & 31;
+	}
+
+	// check if currently in a FX phase, and if so, greatly accelerate phase change.
+	if (SLCN_SPARKLE_FX[self.tmp2] != 0) {
+		sprk_phse_transition_cnt += RNG::Ref().uniform01() * PHASE_THRESHOLD; // good chance of skipping the phase.
```

Comment:
In fact, assuming `sprk_phse_transition_cnt` is uniformly distributed in `[0, PHASE_THRESHOLD]`, the chance that this will shoot it above `PHASE_THRESHOLD` is exactly 1/2.